File: archive/pulumi-ts/Pr1230/lib/components/security.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityInfrastructure = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const aws = __importStar(require("@pulumi/aws"));
class SecurityInfrastructure extends pulumi.ComponentResource {
    webServerSg;
    webServerSgId;
    constructor(name, args, opts) {
        super('tap:components:SecurityInfrastructure', name, args, opts);
        // Create a security group for web servers
        const webServerSgTags = pulumi
            .output(args.tags)
            .apply(t => ({ ...t, Name: `${name}-web-server-sg` }));
        this.webServerSg = new aws.ec2.SecurityGroup(`${name}-web-server-sg`, {
            vpcId: args.vpcId,
            description: 'Allow inbound traffic on port 80 and 443',
            ingress: [
                {
                    protocol: 'tcp',
                    fromPort: 80,
                    toPort: 80,
                    cidrBlocks: ['0.0.0.0/0'],
                    description: 'HTTP from anywhere',
                },
                {
                    protocol: 'tcp',
                    fromPort: 443,
                    toPort: 443,
                    cidrBlocks: ['0.0.0.0/0'],
                    description: 'HTTPS from anywhere',
                },
            ],
            egress: [
                {
                    protocol: '-1',
                    fromPort: 0,
                    toPort: 0,
                    cidrBlocks: ['0.0.0.0/0'],
                    description: 'Allow all outbound traffic',
                },
            ],
            tags: webServerSgTags,
        }, { parent: this });
        // Export key outputs
        this.webServerSgId = this.webServerSg.id;
        this.registerOutputs({
            webServerSgId: this.webServerSgId,
        });
    }
}
exports.SecurityInfrastructure = SecurityInfrastructure;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoic2VjdXJpdHkuanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyJzZWN1cml0eS50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOzs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7QUFBQSx1REFBeUM7QUFDekMsaURBQW1DO0FBU25DLE1BQWEsc0JBQXVCLFNBQVEsTUFBTSxDQUFDLGlCQUFpQjtJQUNsRCxXQUFXLENBQXdCO0lBQ25DLGFBQWEsQ0FBd0I7SUFFckQsWUFDRSxJQUFZLEVBQ1osSUFBZ0MsRUFDaEMsSUFBc0M7UUFFdEMsS0FBSyxDQUFDLHVDQUF1QyxFQUFFLElBQUksRUFBRSxJQUFJLEVBQUUsSUFBSSxDQUFDLENBQUM7UUFFakUsMENBQTBDO1FBQzFDLE1BQU0sZUFBZSxHQUFHLE1BQU07YUFDM0IsTUFBTSxDQUFDLElBQUksQ0FBQyxJQUFJLENBQUM7YUFDakIsS0FBSyxDQUFDLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLEdBQUcsQ0FBQyxFQUFFLElBQUksRUFBRSxHQUFHLElBQUksZ0JBQWdCLEVBQUUsQ0FBQyxDQUFDLENBQUM7UUFDekQsSUFBSSxDQUFDLFdBQVcsR0FBRyxJQUFJLEdBQUcsQ0FBQyxHQUFHLENBQUMsYUFBYSxDQUMxQyxHQUFHLElBQUksZ0JBQWdCLEVBQ3ZCO1lBQ0UsS0FBSyxFQUFFLElBQUksQ0FBQyxLQUFLO1lBQ2pCLFdBQVcsRUFBRSwwQ0FBMEM7WUFDdkQsT0FBTyxFQUFFO2dCQUNQO29CQUNFLFFBQVEsRUFBRSxLQUFLO29CQUNmLFFBQVEsRUFBRSxFQUFFO29CQUNaLE1BQU0sRUFBRSxFQUFFO29CQUNWLFVBQVUsRUFBRSxDQUFDLFdBQVcsQ0FBQztvQkFDekIsV0FBVyxFQUFFLG9CQUFvQjtpQkFDbEM7Z0JBQ0Q7b0JBQ0UsUUFBUSxFQUFFLEtBQUs7b0JBQ2YsUUFBUSxFQUFFLEdBQUc7b0JBQ2IsTUFBTSxFQUFFLEdBQUc7b0JBQ1gsVUFBVSxFQUFFLENBQUMsV0FBVyxDQUFDO29CQUN6QixXQUFXLEVBQUUscUJBQXFCO2lCQUNuQzthQUNGO1lBQ0QsTUFBTSxFQUFFO2dCQUNOO29CQUNFLFFBQVEsRUFBRSxJQUFJO29CQUNkLFFBQVEsRUFBRSxDQUFDO29CQUNYLE1BQU0sRUFBRSxDQUFDO29CQUNULFVBQVUsRUFBRSxDQUFDLFdBQVcsQ0FBQztvQkFDekIsV0FBVyxFQUFFLDRCQUE0QjtpQkFDMUM7YUFDRjtZQUNELElBQUksRUFBRSxlQUFlO1NBQ3RCLEVBQ0QsRUFBRSxNQUFNLEVBQUUsSUFBSSxFQUFFLENBQ2pCLENBQUM7UUFFRixxQkFBcUI7UUFDckIsSUFBSSxDQUFDLGFBQWEsR0FBRyxJQUFJLENBQUMsV0FBVyxDQUFDLEVBQUUsQ0FBQztRQUN6QyxJQUFJLENBQUMsZUFBZSxDQUFDO1lBQ25CLGFBQWEsRUFBRSxJQUFJLENBQUMsYUFBYTtTQUNsQyxDQUFDLENBQUM7SUFDTCxDQUFDO0NBQ0Y7QUF4REQsd0RBd0RDIiwic291cmNlc0NvbnRlbnQiOlsiaW1wb3J0ICogYXMgcHVsdW1pIGZyb20gJ0BwdWx1bWkvcHVsdW1pJztcbmltcG9ydCAqIGFzIGF3cyBmcm9tICdAcHVsdW1pL2F3cyc7XG5cbi8vIERlZmluZSB0aGUgYXJndW1lbnRzIGZvciB0aGUgU2VjdXJpdHlJbmZyYXN0cnVjdHVyZSBjb21wb25lbnRcbmludGVyZmFjZSBTZWN1cml0eUluZnJhc3RydWN0dXJlQXJncyB7XG4gIHZwY0lkOiBwdWx1bWkuSW5wdXQ8c3RyaW5nPjtcbiAgZW52aXJvbm1lbnQ6IHB1bHVtaS5JbnB1dDxzdHJpbmc+O1xuICB0YWdzPzogcHVsdW1pLklucHV0PHsgW2tleTogc3RyaW5nXTogc3RyaW5nIH0+O1xufVxuXG5leHBvcnQgY2xhc3MgU2VjdXJpdHlJbmZyYXN0cnVjdHVyZSBleHRlbmRzIHB1bHVtaS5Db21wb25lbnRSZXNvdXJjZSB7XG4gIHB1YmxpYyByZWFkb25seSB3ZWJTZXJ2ZXJTZzogYXdzLmVjMi5TZWN1cml0eUdyb3VwO1xuICBwdWJsaWMgcmVhZG9ubHkgd2ViU2VydmVyU2dJZDogcHVsdW1pLk91dHB1dDxzdHJpbmc+O1xuXG4gIGNvbnN0cnVjdG9yKFxuICAgIG5hbWU6IHN0cmluZyxcbiAgICBhcmdzOiBTZWN1cml0eUluZnJhc3RydWN0dXJlQXJncyxcbiAgICBvcHRzPzogcHVsdW1pLkNvbXBvbmVudFJlc291cmNlT3B0aW9uc1xuICApIHtcbiAgICBzdXBlcigndGFwOmNvbXBvbmVudHM6U2VjdXJpdHlJbmZyYXN0cnVjdHVyZScsIG5hbWUsIGFyZ3MsIG9wdHMpO1xuXG4gICAgLy8gQ3JlYXRlIGEgc2VjdXJpdHkgZ3JvdXAgZm9yIHdlYiBzZXJ2ZXJzXG4gICAgY29uc3Qgd2ViU2VydmVyU2dUYWdzID0gcHVsdW1pXG4gICAgICAub3V0cHV0KGFyZ3MudGFncylcbiAgICAgIC5hcHBseSh0ID0+ICh7IC4uLnQsIE5hbWU6IGAke25hbWV9LXdlYi1zZXJ2ZXItc2dgIH0pKTtcbiAgICB0aGlzLndlYlNlcnZlclNnID0gbmV3IGF3cy5lYzIuU2VjdXJpdHlHcm91cChcbiAgICAgIGAke25hbWV9LXdlYi1zZXJ2ZXItc2dgLFxuICAgICAge1xuICAgICAgICB2cGNJZDogYXJncy52cGNJZCxcbiAgICAgICAgZGVzY3JpcHRpb246ICdBbGxvdyBpbmJvdW5kIHRyYWZmaWMgb24gcG9ydCA4MCBhbmQgNDQzJyxcbiAgICAgICAgaW5ncmVzczogW1xuICAgICAgICAgIHtcbiAgICAgICAgICAgIHByb3RvY29sOiAndGNwJyxcbiAgICAgICAgICAgIGZyb21Qb3J0OiA4MCxcbiAgICAgICAgICAgIHRvUG9ydDogODAsXG4gICAgICAgICAgICBjaWRyQmxvY2tzOiBbJzAuMC4wLjAvMCddLFxuICAgICAgICAgICAgZGVzY3JpcHRpb246ICdIVFRQIGZyb20gYW55d2hlcmUnLFxuICAgICAgICAgIH0sXG4gICAgICAgICAge1xuICAgICAgICAgICAgcHJvdG9jb2w6ICd0Y3AnLFxuICAgICAgICAgICAgZnJvbVBvcnQ6IDQ0MyxcbiAgICAgICAgICAgIHRvUG9ydDogNDQzLFxuICAgICAgICAgICAgY2lkckJsb2NrczogWycwLjAuMC4wLzAnXSxcbiAgICAgICAgICAgIGRlc2NyaXB0aW9uOiAnSFRUUFMgZnJvbSBhbnl3aGVyZScsXG4gICAgICAgICAgfSxcbiAgICAgICAgXSxcbiAgICAgICAgZWdyZXNzOiBbXG4gICAgICAgICAge1xuICAgICAgICAgICAgcHJvdG9jb2w6ICctMScsXG4gICAgICAgICAgICBmcm9tUG9ydDogMCxcbiAgICAgICAgICAgIHRvUG9ydDogMCxcbiAgICAgICAgICAgIGNpZHJCbG9ja3M6IFsnMC4wLjAuMC8wJ10sXG4gICAgICAgICAgICBkZXNjcmlwdGlvbjogJ0FsbG93IGFsbCBvdXRib3VuZCB0cmFmZmljJyxcbiAgICAgICAgICB9LFxuICAgICAgICBdLFxuICAgICAgICB0YWdzOiB3ZWJTZXJ2ZXJTZ1RhZ3MsXG4gICAgICB9LFxuICAgICAgeyBwYXJlbnQ6IHRoaXMgfVxuICAgICk7XG5cbiAgICAvLyBFeHBvcnQga2V5IG91dHB1dHNcbiAgICB0aGlzLndlYlNlcnZlclNnSWQgPSB0aGlzLndlYlNlcnZlclNnLmlkO1xuICAgIHRoaXMucmVnaXN0ZXJPdXRwdXRzKHtcbiAgICAgIHdlYlNlcnZlclNnSWQ6IHRoaXMud2ViU2VydmVyU2dJZCxcbiAgICB9KTtcbiAgfVxufVxuIl19

File: lib/components/security.py
# components/security.py
"""
Security component that creates security groups, IAM roles, and WAF
Implements least privilege access and defense in depth
"""

import json
import pulumi
import pulumi_aws as aws
VPC_CIDRS = ["10.0.0.0/16", "10.1.0.0/16"]


class SecurityComponent(pulumi.ComponentResource):
  def __init__(self, name: str, vpc_id: pulumi.Output[str], subnets: str, region: str, tags: dict, opts: pulumi.ResourceOptions = None):
    super().__init__("custom:security:SecurityComponent", name, None, opts)

    self.region = region
    self.tags = tags

    # Create KMS key for encryption at rest
    self.kms_key = aws.kms.Key(
        f"{name}-kms-key",
        description="KMS key for encrypting application resources",
        tags=tags,
        opts=pulumi.ResourceOptions(parent=self)
    )

    self.kms_alias = aws.kms.Alias(
        f"{name}-kms-alias",
        name=f"alias/{name}-encryption-key",
        target_key_id=self.kms_key.key_id,
        opts=pulumi.ResourceOptions(parent=self)
    )

    # Security Group for Application Load Balancer
    # Only allows HTTP/HTTPS from internet, redirects HTTP to HTTPS
    self.alb_security_group = aws.ec2.SecurityGroup(
        f"{name}-alb-sg",
        name_prefix=f"{name}-alb-",
        description="Security group for Application Load Balancer",
        vpc_id=vpc_id,
        ingress=[
            # Allow HTTP (will redirect to HTTPS)
            aws.ec2.SecurityGroupIngressArgs(
                from_port=80,
                to_port=80,
                protocol="tcp",
                cidr_blocks=["0.0.0.0/0"],
                description="HTTP from internet"
            ),
            # Allow HTTPS
            aws.ec2.SecurityGroupIngressArgs(
                from_port=443,
                to_port=443,
                protocol="tcp",
                cidr_blocks=["0.0.0.0/0"],
                description="HTTPS from internet"
            )
        ],
        egress=[
            # Allow all outbound to Lambda functions
            aws.ec2.SecurityGroupEgressArgs(
                from_port=0,
                to_port=65535,
                protocol="tcp",
                cidr_blocks=["10.0.0.0/8"],
                description="All traffic to private subnets"
            )
        ],
        tags={**tags, "Name": f"{name}-alb-sg"},
        opts=pulumi.ResourceOptions(parent=self)
    )

    # Security Group for Lambda functions
    # Allows inbound from ALB and outbound to RDS/DynamoDB/S3
    self.lambda_security_group = aws.ec2.SecurityGroup(
        f"{name}-lambda-sg",
        name_prefix=f"{name}-lambda-",
        description="Security group for Lambda functions",
        vpc_id=vpc_id,
        ingress=[
            # Allow traffic from ALB
            aws.ec2.SecurityGroupIngressArgs(
                from_port=80,
                to_port=80,
                protocol="tcp",
                # source_security_group_id=self.alb_security_group.id,
                cidr_blocks=["0.0.0.0/0"],
                description="HTTP from ALB"
            )
        ],
        egress=[
            # Allow HTTPS for AWS API calls (S3, DynamoDB)
            # Allow MySQL/PostgreSQL to RDS
            aws.ec2.SecurityGroupEgressArgs(
                from_port=5432,
                to_port=5432,
                protocol="tcp",
                cidr_blocks=VPC_CIDRS,
                description="PostgreSQL to RDS"
            ),
            # aws.ec2.SecurityGroupEgressArgs(
            #     protocol="icmp",        # ICMP
            #     # ICMP type (8 = Echo Request). Use -1 for all types
            #     from_port=-1,
            #     to_port=-1,             # ICMP code (-1 = all codes)
            #     cidr_blocks=VPC_CIDRS
            # )

        ],
        tags={**tags, "Name": f"{name}-lambda-sg"},
        opts=pulumi.ResourceOptions(parent=self)
    )

    # Security Group for RDS Database
    # Only allows inbound from Lambda functions on database port
    self.database_security_group = aws.ec2.SecurityGroup(
        f"{name}-database-sg",
        name_prefix=f"{name}-database-",
        description="Security group for RDS database",
        vpc_id=vpc_id,
        ingress=[
            # Allow PostgreSQL from Lambda functions
            aws.ec2.SecurityGroupIngressArgs(
                from_port=5432,
                to_port=5432,
                protocol="tcp",
                # source_security_group_id=self.lambda_security_group.id,
                cidr_blocks=VPC_CIDRS,
                description="PostgreSQL from Lambda"
            ),
            aws.ec2.SecurityGroupIngressArgs(
                protocol="icmp",        # ICMP
                # ICMP type (8 = Echo Request). Use -1 for all types
                from_port=-1,
                to_port=-1,             # ICMP code (-1 = all codes)
                cidr_blocks=VPC_CIDRS
            )
        ],
        # No egress rules needed for RDS
        tags={**tags, "Name": f"{name}-database-sg"},
        opts=pulumi.ResourceOptions(parent=self)
    )

    # IAM Role for Lambda execution with least privilege
    lambda_assume_role_policy = {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": {
                    "Service": "lambda.amazonaws.com"
                }
            }
        ]
    }

    self.lambda_execution_role = aws.iam.Role(
        f"{name}-lambda-execution-role",
        assume_role_policy=json.dumps(lambda_assume_role_policy),
        tags=tags,
        opts=pulumi.ResourceOptions(parent=self)
    )

    # Lambda execution policy with minimal required permissions
    json_lambda_policy = {
        "Version": "2012-10-17",
        "Statement": [
            # CloudWatch Logs permissions
            {
                "Effect": "Allow",
                "Action": [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents"
                ],
                "Resource": f"arn:aws:logs:{region}:*:*"
            },
            # VPC permissions for Lambda
            {
                "Effect": "Allow",
                "Action": [
                    "ec2:CreateNetworkInterface",
                    "ec2:DescribeNetworkInterfaces",
                    "ec2:DeleteNetworkInterface",
                    "ec2:AttachNetworkInterface",
                    "ec2:DetachNetworkInterface"
                ],
                "Resource": "*"
            },
            # DynamoDB permissions (scoped to specific table)
            {
                "Effect": "Allow",
                "Action": [
                    "dynamodb:GetItem",
                    "dynamodb:PutItem",
                    "dynamodb:UpdateItem",
                    "dynamodb:DeleteItem",
                    "dynamodb:Query",
                    "dynamodb:Scan"
                ],
                "Resource": f"arn:aws:dynamodb:{region}:*:table/pulumi-optimization-*"
            },
            # S3 permissions (scoped to specific bucket)
            {
                "Effect": "Allow",
                "Action": [
                    "s3:GetObject",
                    "s3:PutObject",
                    "s3:DeleteObject"
                ],
                "Resource": f"arn:aws:s3:::pulumi-optimization-*/*"
            },
            # KMS permissions for decryption
            {
                "Effect": "Allow",
                "Action": [
                    "kms:Decrypt",
                    "kms:GenerateDataKey"
                ],
                "Resource": "*"
            }
        ]
    }

    self.lambda_policy = aws.iam.RolePolicy(
        f"{name}-lambda-policy",
        role=self.lambda_execution_role.id,
        policy=json.dumps(json_lambda_policy),
        opts=pulumi.ResourceOptions(parent=self)
    )

    # LOCALSTACK FIX: WAF removed (Pro-only feature)
    # LocalStack Community Edition does not support WAF
    # self.waf_web_acl = aws.wafv2.WebAcl(...)

    self.alb = aws.lb.LoadBalancer(
        f"{name}-alb",
        security_groups=[self.alb_security_group.id],
        subnets=subnets,
        load_balancer_type="application",
        opts=pulumi.ResourceOptions(parent=self),
    )

    self.register_outputs({})
